storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3109,1764 @@ buf_block_t *fsp_get_latched_xdes_page(
+    could've changed while allocating the new block. */
+    pageRemover remover(header_block, iblock, inode, extent_size,
+                        cur_page_no, mtr);
+    err= remover.prepare();
+ 
+    DBUG_EXECUTE_IF("remover_prepare_fail", err= DB_CORRUPTION;);
+    if (err)
+    {
+      allocator.restore_old_pages();
+      mtr->discard_modifications();
+      goto err_exit;
+    }
+
+    /* Copy the data from old block to new block */
+    buf_block_t *new_block= allocator.get_new_block();
+    uint32_t new_page_no= new_block->page.id().page_no();
+    /* Copy FIL_PAGE_PREV, FIL_PAGE_NEXT */
+    mtr->memcpy(*new_block, new_block->page.frame + FIL_PAGE_PREV,
+                block->page.frame + FIL_PAGE_PREV, 8);
+    mtr->memcpy(*new_block, new_block->page.frame + FIL_PAGE_TYPE,
+                block->page.frame + FIL_PAGE_TYPE,
+		srv_page_size - FIL_PAGE_TYPE - 8);
+
+    /* Assign the new block page number in left, right
+    and parent block */
+    related_pages.complete(new_page_no, parent_offset);
+
+    /* Complete the page free operation */
+    remover.complete(); 
+    /* Add the new page in inode fragment array */
+    allocator.assign_frag_slot();
+
+    if (level)
+    {
+      err= store_child_pages(new_block);
+      if (err) goto err_exit;
+    }
+    goto fetch_next_page;
+  }
+
+  ut_ad(!fsp_tablespace_validate(space));
+  ut_ad(!fseg_validate(space, m_index));
+  root->page.unfix();
+  return DB_SUCCESS;
+}
+
+dberr_t indexDefragmenter::defragment(spaceDefragmenter *space_defrag) noexcept
+{
+  mtr_t mtr;
+  dberr_t err= DB_SUCCESS;
+  m_index->lock.x_lock(SRW_LOCK_CALL);
+  fil_space_t *const space= fil_system.sys_space;
+  mtr.start();
+  mtr.x_lock_space(space);
+  buf_block_t *root= fsp_get_latched_page(page_id_t{0, m_index->page},
+                                          &mtr, &err);
+  if (!root)
+  {
+    mtr.commit();
+    m_index->lock.x_unlock();
+    return err;
+  }
+  uint16_t level= btr_page_get_level(root->page.frame);
+  while (1)
+  {
+    err= defragment_level(level, &mtr, space_defrag);
+    DBUG_EXECUTE_IF("fail_after_level_defragment",
+                    if (level == 1)
+		      err= DB_CORRUPTION;);
+    if (err || !level)
+      break;
+    level--;
+    mtr.start();
+    mtr.x_lock_space(space);
+  }
+  m_index->lock.x_unlock();
+  return err;
+}
+
+static
+bool is_user_table_exists(fil_space_t *space, dberr_t *err) noexcept
+{
+  mtr_t mtr;
+  btr_pcur_t pcur;
+  mtr.start();
+  for (const rec_t *rec= dict_startscan_system(&pcur, &mtr,
+			                       dict_sys.sys_tables);
+       rec; rec= dict_getnext_system(&pcur, &mtr))
+  {
+    if (rec_get_deleted_flag(rec, 0))
+    {
+      sql_print_error("InnoDB: Encountered corrupted record in SYS_TABLES");
+      *err= DB_CORRUPTION;
+user_table_found:
+      mtr.commit();
+      return true;
+    }
+    ulint len;
+    const byte *field=
+      rec_get_nth_field_old(rec, DICT_FLD__SYS_TABLES__SPACE, &len);
+    ut_ad(len == 4);
```

Comment:
Assertions must be replaced with a `goto` to the corruption reporting. Would the API be simpler if we returned `DB_SUCCESS` or `DB_SUCCESS_LOCKED_REC` on not-found/found and another error value on corruption? In that way, the caller would not have to allocate stack for an output parameter.